modules/standard/CTypes.chpl

```diff
@@ -836,16 +836,16 @@ module CTypes {
   */
   @chpldoc.nodoc
   inline proc c_ptrTo(ref arr: []): c_ptr(arr.eltType) {
-    if (!arr.isRectangular() || !arr.domain.distribution._value.dsiIsLayout()) then
```

Comment:
Though I don't have a preference between the two ways of writing this, I also don't understand what wasn't static about the old version -- it only used param intent procs.

Reply:
I don't have a satisfying explanation but this has to do with runtime types.

```chpl
proc foo(ref arr: []) {
  if (!arr.isRectangular() || !arr.domain.distribution._value.dsiIsLayout()) {
    compilerWarning("in foo");
  }

  return arr[2];
}

proc bar(ref arr: []) {
  if (!isSubtype(arr.domain._instance.type, DefaultRectangularDom)) {
    compilerWarning("in bar");
  }

  return arr[2];
}

var arr = [1,2,3];
writeln(foo(arr));
writeln(bar(arr));
```

Here, in `foo`, accessing the distribution of the domain is left in the AST, codegenned, and run at execution time. I believe this is because the compiler being too skittish to remove them, again, likely due to runtime types.

OTOH, in `bar`, I see the `if` being completely removed from the AST.

This is not a very satifying explanation as it doesn't cover the fact that domains also have runtime type components and both `foo` and `bar` use `arr.domain`. But that's the best I got before digging further.